### ISSUE TITLE
Action & EventsRef

### ISSUE THREAD

**Issue by jorgenj** [serverlessworkflow/specification#151]
**What would you like to be added**:

The EventsRef option for [Action definition](https://github.com/serverlessworkflow/specification/blob/master/specification.md#action-definition) is confusing, it really seems to be an RPC call where the transport protocol is cloud-events (the remote-procedure is invoked via an event and the response is received via another event).  This really seems like another way to invoke a 'function', and thus, in my humble opinion, should be redefined as a type of function, rather than a separate type of action.

**Why is this needed**:

As the Action is defined now, it's really unclear at first read what these 'events' in the eventsRef are for.  It took me several reads through the entire spec to understand what this is doing.  It would be much easier to understand if this were redefined so that this is just another type of 'function' rather than a different type of action.

Proposal:

* Remove the 'eventsRef' attribute from 'action' definition
* Add support for a new type of 'function', re-purposing the description for eventsRef.

Same behavior, but as a 'function', because what it seems to be describing is a way to 'invoke' something via cloud-event.

**Comment by tsurdilo** [comment]
@jorgenj Sorry I respectfully disagree with this :) 
Maybe the documentation is not clear or does not give enough examples. I give you that for sure.
However creating a "function" definition for this is imo not the best approach and the one we have now IMO does what is expected which is to produce an event and then wait for a results event to arrive before proceeding workflow exec.

There are many situations where you want to invoke 1 or more services/functions and where those are triggered by events, see for example https://quarkus.io/guides/funqy. To support these types of situations we only need to worry about sending the appropriate event, and then wait for an event which describes the completion of 1+ services/functions that are invoked after.

Using trigger and result ref event references in this case to me at least makes perfect sense as it is in the spec currently.

**Comment by tsurdilo** [comment]
@jorgenj just to add sorry forgot, when you use "function" definitions, you know exactly what service you want to invoke and what operation on it you want to trigger. With event-based actions you might not know or need to know what service will be invioked or how many, that is the use case of it. Hope this helps.

**Comment by jorgenj** [comment]
Following up from the discussion on the call this morning, there was some good discussion on this idea.  My take on the discussion (correct me if I'm wrong) was that there is support for exploring this idea further.  

Here's an example workflow that would use eventRef-based action, but re-imagined as another type of 'function' rather than a different type of action:

```
{  
   "id": "sampleWorkflow",
   "version": "1.0",
   "name": "Sample Workflow",
   "description": "Sample Workflow",
   "events": [
    {
      "names": ["StartEvent", "DoneEvent"],
      "type": "org.events",
      "source": "eventssource"
    }
  ],
   "functions": [
       {
          "name": "FnFoo",
          "type": "event",
          "triggerEventRef": "StartEvent",
          "resultEventRef":  "DoneEvent",
          ...
       },
       {  
          "name": "FnBar",
          "operation": "https://hellworldservice.test.com:8443/api/api.yaml#myOperation",
          ...
       }
   ],
   "states":[
      {
        "name": "MyState",
        "type": "operation",
        "actionMode": "sequential",
        "actions": [
            {
                "refName": "FnFoo",
                "parameters": {
                    "applicant": "{{ $.customer }}"
                }
            },
            {
                "refName": "FnBar",
                "parameters": {
                    "applicant": "{{ $.customer }}"
                }
            },
        ],
        "end": {
            "kind": "default"
        }
    }
   ]
}
```

Note that, for this example I assumed that the properties of 'functionRef' under 'action' would be merged into the action definition, and that the eventRef and functionRef attributes would then be removed.

**Comment by jorgenj** [comment]
@tsurdilo @manuelstein What do you think about the example proposed above?

**Comment by tsurdilo** [comment]
@jorgenj couple things that come to mind looking at the provided example:

1) I rather not make function and event definitions depend on each other, especially since we claim they are reusable. In your example the runtimes pretty much are forced to parse the events definitions first. We added recently ability to externalize function and events defs into separate files, and adding a dependency now between these two arrays will bring more problems then solve them.

2) the use of type=event is clunky imo, we also removed the type param from function defs recently.

3) We currently have action data filters and event filters (which needs to be revised anyways). It is not clear which to use at this point as with function exec we have the function results, with events we have the event body as result. So this would complicate things some more around that imo.

4) "parameters" makes sense for functions, but not for events, where we are dealing with context properties and event body data.

I will take a deeper look but this is just a quick overview.
WDYT?

**Comment by manuelstein** [comment]
Hi, very good points!
Regarding 1, I think it could be solved allowing inline event definitions instead of eventRefs. If the subroutine consists of sending an event of type X and expecting an event type Y from a specific source, then that's fine.
```yaml
functions: 
  - name: "FnFoo"
    type: "event"
    outgoing:
      specversion: "1.0"
      type: "com.example.order.accepted"
      source: "/users/should/control/that"
      id: "should:probably.be-generated@by_workflowengine"
      dataschema: "http://my.schema.com/outgoing.json"
      data: |
        {
          "maybe": "we",
          "can": "allow",
          "defaults": "and",
          "references": [
            "$my.path.to.workflowdata",
            "$another.path.to.data",
            "$replace.me.magically".
            "$should_these_become_parameternames"
          ]
        }
    incoming:
    - type: "com.example.goods.delivered"
      source: "/from/the/delivery/team"
states:
  - name: "MyState"
    type: "operation"
    actionMode: "sequential"
    actions:
    - refName: "FnFoo"
      parameters:
        - should_these_become_parameternames: "$.nowWereTalking"
```

But looking at the example really makes me wonder if we want to wrap event exchanges with function declarations.
Is sending and then waiting on an event so much more difficult?
It would also allow us to send it in some state and consume events that result from this trigger much later.

About the `type = event`, I thought we had made `type = openapi` the default so it can be omitted.

I think the filters in eventDataFilter and actionDataFilter do the same thing, only that action can have filters on input and output data, event can only have output filters, but that is kind of confusing when it applies to an incoming event, so yes, maybe this can be aligned. I'm not sure if the filtering the data that can be used as input parameters is a useful thing. It allows the workflow designer to filter what can be used, and as the next step the workflow designer would select what is being used.
If it was all ouputFilters, i.e. decide which paths from the result go where into the JSON workflow data, then a single filter mechanism for all incoming JSON structured data would be enough.
The exception may be states, where filters on outgoing workflow data may be useful, but only from a runtime implementation point of view, not from the designer's perspective.

Good discussion, I'm also going to take another look.

**Comment by manuelstein** [comment]
In summary of today's discussion (please add points I'm missing):

- the proposed type=event function definitions would make it easier for users to make event-based "invocations" (situations where an event is sent and some other event is awaited as response)
- reusing existing outgoing/incoming event definitions should be the baseline functionality; inline definitions could be an additional feature
- the event definition must not reference runtime workflow data (references into future workflow data makes it even less reusable)

also noted:
- a function call is a transaction per HTTP transport, so results and errors invoking the function should be visible in responses, i.e. within the scope of the transaction; but an event-based invocation may have other independent outcomes signalling an error or result; regarding errors, the timeout to send the outgoing event as well as a timeout to receive the expected response would be the only error cases for now; regarding multiple "reponse" events may require some more design

**Comment by manuelstein** [comment]
Another idea that came offline is:
Making an event-based invocation a "function call" still feels weird. The spec currently defines actions and the only actions we're putting there are function calls. An event-based invocation or trigger of work is definitely an action, so we need to somehow have an action that sends an event.
Thx @tsurdilo for pointing out that we have one kind of [event-based service invocation example](https://github.com/serverlessworkflow/specification/blob/master/examples/examples.md#event-based-service-invocation) in which the state action invokes work with a function call (asynchronous) and then waits on an eventRef. What we would need here is to have an action that just sends an event, e.g. an outgoing event definition as an action (and selecting some metadata + data)

**Comment by tsurdilo** [comment]
@manuelstein nice, thanks for the summary!! Will think about it some more and try to have more info for our next meet. Many thanks.

**Comment by tsurdilo** [comment]
After thinking some more about this I feel that 
1) we need to keep event and function defs separate. no references or some sort of combinations
2) if improvements need to be done then those should be in the actions definitions
